Build is angry:

>     /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang -x objective-c++ -ivfsstatcache /Users/buildbot/Library/Developer/Xcode/DerivedData/SDKStatCaches.noindex/iphoneos18.5-22F76-7fa4eea80a99bbfdc046826b63ec4baf.sdkstatcache -fmessage-length\=0 -fdiagnostics-show-note-include-stack -fmacro-backtrace-limit\=0 -fno-color-diagnostics -Wno-trigraphs -Werror -Wno-missing-field-initializers -Wmissing-prototypes -Wunreachable-code -Wno-implicit-atomic-properties -Wno-objc-interface-ivars -Wno-arc-repeated-use-of-weak -Wimplicit-retain-self -Wnon-virtual-dtor -Wno-overloaded-virtual -Wno-exit-time-destructors -Wduplicate-method-match -Wno-missing-braces -Wparentheses -Wswitch -Wunused-function -Wno-unused-label -Wno-unused-parameter -Wunused-variable -Wunused-value -Wempty-body -Wuninitialized -Wno-unknown-pragmas -Wno-shadow -Wno-four-char-constants -Wno-conversion -Wconstant-conversion -Wint-conversion -Wbool-conversion -Wenum-conversion -Wno-float-conversion -Wnon-literal-null-conversion -Wobjc-literal-conversion -Wsign-compare -Wno-shorten-64-to-32 -Wnewline-eof -Wno-selector -Wno-strict-selector-match -Wundeclared-selector -Wdeprecated-implementations -Wno-c++11-extensions -Wno-implicit-fallthrough -fstrict-aliasing -Wprotocol -Wdeprecated-declarations -Winvalid-offsetof -Wno-sign-conversion -Winfinite-recursion -Wmove -Wcomma -Wblock-capture-autoreleasing -Wstrict-prototypes -Wrange-loop-analysis -Wno-semicolon-before-method-body -Wall -Wc99-designator -Wconditional-uninitialized -Wextra -Wdeprecated-enum-enum-conversion -Wdeprecated-enum-float-conversion -Wenum-float-conversion -Wfinal-dtor-non-final-class -Wformat\=2 -Wmisleading-indentation -Wpointer-to-int-cast -Wreorder-init-list -Wundef -Wunused-but-set-variable -Wvla -Wcast-qual -Wchar-subscripts -Wextra-tokens -Winit-self -Wmissing-noreturn -Wpacked -Wpointer-arith -Wredundant-decls -Wwrite-strings -Wexit-time-destructors -Wglobal-constructors -Wtautological-compare -Wimplicit-fallthrough -Wvla -Wliteral-conversion -Wthread-safety -Wno-typedef-redefinition -Wunsafe-buffer-usage @/Volumes/Data/worker/iOS-18-Build-EWS/build/WebKitBuild/WebGPU.build/Release-iphoneos/WebGPU.build/Objects-normal/arm64/af3fcb34312c57c0f52879cdce924b91-common-args.resp -MMD -MT dependencies -MF /Volumes/Data/worker/iOS-18-Build-EWS/build/WebKitBuild/WebGPU.build/Release-iphoneos/WebGPU.build/Objects-normal/arm64/HardwareCapabilities.d --serialize-diagnostics /Volumes/Data/worker/iOS-18-Build-EWS/build/WebKitBuild/WebGPU.build/Release-iphoneos/WebGPU.build/Objects-normal/arm64/HardwareCapabilities.dia -c /Volumes/Data/worker/iOS-18-Build-EWS/build/Source/WebGPU/WebGPU/HardwareCapabilities.mm -o /Volumes/Data/worker/iOS-18-Build-EWS/build/WebKitBuild/WebGPU.build/Release-iphoneos/WebGPU.build/Objects-normal/arm64/HardwareCapabilities.o
> In file included from /Volumes/Data/worker/iOS-18-Build-EWS/build/Source/WebGPU/WebGPU/HardwareCapabilities.mm:29:
> /Volumes/Data/worker/iOS-18-Build-EWS/build/Source/WebGPU/WebGPU/Internal/WebGPUSwiftInternal.h:144:10: fatal error: 'WebGPUSwift-Generated.h' file not found
>   144 | #include "WebGPUSwift-Generated.h"
>       |          ^~~~~~~~~~~~~~~~~~~~~~~~~
> 1 error generated.

I don't see anything in the build output that explains why this file was not generated though 🥺.

--- a/Source/WebGPU/WebGPU/HardwareCapabilities.mm
+++ b/Source/WebGPU/WebGPU/HardwareCapabilities.mm
@@ -167,17 +167,74 @@ bool isShaderValidationEnabled(id<MTLDevice> device)
     return result;
 }
 
+#if ENABLE(WEBGPU_SWIFT)
+static std::optional<std::array<int, 4>> extractClangVersion()
+{
+    NSString* input = @(__clang_version__);
+    NSError *error = nil;
+    NSRegularExpression *regex = [NSRegularExpression
+        regularExpressionWithPattern:@"clang-([0-9]+)\\.([0-9]+)\\.([0-9]+)\\.([0-9]+)"
+        options:0
+        error:&error];
+
+    if (error)
+        return std::nullopt;
+
+    NSTextCheckingResult *match = [regex firstMatchInString:input options:0 range:NSMakeRange(0, input.length)];
+
+    if (match && match.numberOfRanges > 1) {
+        std::array<int, 4> result;
+        for (int i = 0; i < 4; ++i) {
+            NSRange clangRange = [match rangeAtIndex:i + 1];
+            NSString* substring = [input substringWithRange:clangRange];
+            result[i] = substring.intValue;
+        }
+        return result;
+    }
+
+    return std::nullopt;
+}
+
+static bool swiftCompilerSupportsWebGPU()
+{
+    auto maybeClangVersion = extractClangVersion();
+    if (!maybeClangVersion)
+        return false;
+
+    auto clangVersion = *maybeClangVersion;
+    if (clangVersion[0] == 1700 && clangVersion[1] >= 6 && clangVersion[2] >= 1 && clangVersion[3] >= 1)
+        return true;
+
+    if (clangVersion[0] == 2100 && clangVersion[1] >= 0 && clangVersion[2] >= 101 && clangVersion[3] >= 15)
+        return true;
+
+    if (clangVersion[0] > 2100)
+        return true;
+
+    return false;
+}
+#endif
+
 bool isWebGPUSwiftEnabled()
 {
 #if defined(ENABLE_LIBFUZZER) && ENABLE_LIBFUZZER && defined(ASAN_ENABLED) && ASAN_ENABLED
     return true;
+#elif !ENABLE(WEBGPU_SWIFT)
+    return false;
 #else
     static std::once_flag onceFlag;
     static bool isWebGPUSwiftEnabled;
     std::call_once(onceFlag, [&] {
-        isWebGPUSwiftEnabled = [[NSUserDefaults standardUserDefaults] boolForKey:@"WebKitWebGPUSwiftEnabled"];
+        NSNumber* object = [[NSUserDefaults standardUserDefaults] objectForKey:@"WebKitWebGPUSwiftEnabled"];
+        if (object)
+            isWebGPUSwiftEnabled = object.boolValue;
+        else
+            isWebGPUSwiftEnabled = swiftCompilerSupportsWebGPU();
+
         if (isWebGPUSwiftEnabled)
             WTFLogAlways("WebGPU: using SWIFT backend"); // NOLINT
+        else
+            WTFLogAlways("WebGPU: using C++ backend"); // NOLINT
     });
     return isWebGPUSwiftEnabled;
 #endif

